Added example codes with images only and pics by webcam. Updated Readme.

rounaksingh opened the pull request:
It's working good!

Thanks for porting zbar :)

rounaksingh committed:
Updated Readme.md. updated zbar.py. Added testing code and test barcodes

rounaksingh committed:
Updated Readme

rounaksingh committed:
Testing with webcam added

rounaksingh committed:
Updated Readme

rounaksingh committed:
Test of zbar, cam and finding food product name is added

zpincus commented:
Thanks for the tests and updates! This will be very useful!

I'll make a few comments on the code shortly -- there are a couple things that could be handled a little differently.

zpincus commented on .gitignore:
```diff
@@ -55,3 +55,9 @@ docs/_build/
 # PyBuilder
 target/
+
+# Image files
```
See comments on other files: probably better to not generate these files in the first place (or clean them up in the code) rather than asking git to ignore them...

rounaksingh replied:
This is just for short duration. I am working to get image array from pygame directly.

zpincus commented on zbar_testing/1_Testing_barcode_images/zbar_test.py:
```diff
@@ -0,0 +1,46 @@
+# Test for zbar-py
```
This and the other "tests" are more like "examples". If you want to add actual tests, it would be a good idea to do so with the [unittest](https://docs.python.org/3/library/unittest.html) framework.

Also, for tests, you could avoid the dependency on `scipy.misc.imread` by loading and then saving out the image files as `.npy` file with `numpy.save()`, and then in the test code, using `numpy.load` to read in the data. This isn't super realistic for an example, but it's great for a test.

I suggest moving the current "tests" to an "examples" directory, and then if you want to, you could rework the first example as a set of test cases, using all of the different images.

rounaksingh replied:
I will change the directory name to examples.

rounaksingh replied:
Actually when I first start testing zbar-py (how it works), I didnot know that time that all my testing will become examples.

zpincus commented on zbar_testing/2_Testing_barcode_cam/zbar_test.py:
```diff
@@ -0,0 +1,88 @@
+# Test for zbar-py using a webcam
```
What's the deal with the `2_Testing_barcode_cam/test_images_by_cam/*.jpg` files? They're not used in the the code below, so why are they checked into git?

Also, it's a little inelegant to save and then reload the image files. Why not just convert them from pygame surfaces to numpy arrays directly, [as follows](http://www.petercollingridge.co.uk/book/export/html/550):
```python
def getPixelArray(filename):
    image = pygame.image.load(filename)
    return pygame.surfarray.array3d(image)
```

rounaksingh replied:
I am working on getting image array directly from pygame. I will update it.

zpincus commented on zbar_testing/3_zbar_cam_productName/zbar_GetProductName.py:
```diff
@@ -0,0 +1,127 @@
+# Test for zbar-py using a webcam
```
Could probably just roll this into the second example. Otherwise it's a lot of duplicated code...

rounaksingh committed:
Scanner return value for barcode data is changed from bytearray to unicode str

rounaksingh committed:
New file: zbar.misc added. It contains functions for barcode conversion and checksum calculation

rounaksingh committed:
setup.py updated to work with setuptools

zpincus commented on setup.py:
```diff
@@ -1,7 +1,5 @@
-import distutils.core
-import distutils.errors
+from setuptools import setup
+from setuptools.extension import Extension
```
Ideally the script would still work if setuptools (which isn't part of the standard library yet) isn't available. It should be as simple as trying to import setuptools and if not present, falling back on the distutils functions of the same name...

rounaksingh replied:
I agree that the setuptools are not part of standard library yet.

* Because of the [article](http://stackoverflow.com/questions/6344076/differences-between-distribute-distutils-setuptools-and-distutils2). I think it's good to use setuptools. Though, I will add a falling back in next commit.

* There are some setuptools.core.setup() options like install_requires which do not work with distutils.core.setup(). I think it is better to have dependencies to be installed.

zpincus commented on zbar/zbar.py:
```diff
@@ -171,7 +171,7 @@ def scan(self, image):
             sym_data_ptr = _ZB.zbar_symbol_get_data(symbol)
             sym_data_len = _ZB.zbar_symbol_get_data_length(symbol)
-            sym_data = ctypes.string_at(sym_data_ptr, sym_data_len)
+            sym_data = ctypes.string_at(sym_data_ptr, sym_data_len).decode("utf-8")
```
There's no guarantee that an arbitrary barcode's data will be text encoded as UTF-8. It could be raw binary data, or text in some other encoding. So if a client of zbar knows the data from their barcode is UTF-8 encoded, they should call decode, but this library shouldn't be making that assumption.

Indeed, the docstring for `scan()` specifically states: "data' is a bytes instance containing the barcode payload", so this counts as a non-backward-compatible API change.

For those reasons, I think that this change should be reconsidered.

zpincus commented:
Thanks again for these examples and utility code! Hopefully the comments above are helpful and seem reasonable.

rounaksingh committed:
Class Scanner:new members added scan_from_image(). Release moved to 1.2rc1

rounaksingh committed:
README and zbar_testing updated.

zpincus commented on setup.py:
```diff
@@ -53,14 +53,14 @@ def has_libc_iconv():
         ext_modules=[zbar],
         packages=['zbar'],
-        install_requires=['numpy'],
+        install_requires=['numpy','scipy'],
```
Scipy is a very heavyweight requirement, especially since it's just for image reading -- which is not even a core part of the zbar functionality. There are **lots** of ways to get images into numpy arrays; using the most appropriate for any specific application should be the zbar user's responsibility. By all means have the scipy-based image reading as example code, though!

zpincus commented on setup.py:
```diff
@@ -53,14 +53,14 @@ def has_libc_iconv():
+setup(name='zbar-py',
+        version='1.2rc1',
         description='zbar package',
         url='https://github.com/rounaksingh/zbar-py',
```
You don't think the canonical URL should be at the zplab github page?

zpincus commented on zbar/zbar.py:
```diff
@@ -181,4 +182,32 @@ def scan(self, image):
\ No newline at end of file
+        return symbols
+
+    def rgb2gray(self,rgb):
```
`rgb2gray` probably belongs in `zbar.misc`, or just as an example.

rounaksingh replied:
you are correct. It should go to zbar.misc

zpincus commented on zbar/zbar.py:
```diff
@@ -181,4 +182,32 @@ def scan(self, image):
+        '''
+        return numpy.dot(rgb[...,:3], [0.299, 0.587, 0.114])
+
+    def scan_from_image(self,image_filename):
```
As in the comment to `setup.py`, this is probably better as an example than a core part of the library, if only to avoid the scipy dependency.

rounaksingh committed:
testing directory name changed to examples

rounaksingh committed:
reading images is removed from core functionality. Scipy dependency removed

rounaksingh committed:
setup.py changed back to disutils

rounaksingh committed:
zbar.scan() barcode data return type changed to bytearray

rounaksingh committed:
Examples and README updated

rounaksingh committed:
Updated Readme.md

rounaksingh commented:
Hi I have changed things. Please review.

rounaksingh commented:
* Removed dependency on scipy. For examples only pygame required.
* setup.py links were just for test purposes, I have changed them.
* setup.py is using distutils
* scan from image function is removed from zbar core.
* examples now donot create image file. Images are loaded and converted into ndarray directly
* rgb2gray() is shifted to zbar.misc
* .gitignore: Images name updated

If there is something else, let me know.
Also, can you please make it available in PyPI.
